core/state-db/src/pruning.rs

```diff
@@ -143,13 +141,6 @@ impl<BlockHash: Hash, Key: Hash> RefWindow<BlockHash, Key> {
 			commit.meta.inserted.push((to_meta_key(LAST_PRUNED, &()), index.encode()));
 			commit.meta.deleted.push(pruned.journal_key.clone());
 			self.pending_prunings += 1;
-		} else if let Some((block, pruned)) = self.pending_records.get(self.pending_prunings - self.death_rows.len()) {
```

Comment:
Pending finalizations end up in `self.deathrows` now. So the block above can now prune both existing and pending blocks. `revert_pending` should handle it just fine.

Reply:
`revert_pending` does not rollback those data sent to the CommitSet so if we do not send them in the first place. But it is also likely to break some loop exit condition.

Reply:
`revert_pending` is supposed to be called when the `CommitSet` did not end up in the DB because of some failure. The goal here is to keep internal state consistent with the database in case this happens. `revert_pending` basically tells us that all data that we returned in previous calls to `note_canonical` and `prune_one` was lost.

Reply:
Yes that makes sense, something to keep in mind :+1:  (still think explicitly making it not possible could make sense for additional safety and avoid future regression, but actually it may bring some regression right now so I understand it may not be worth it :)